### ISSUE TITLE
Proposal: Use OAuth to authorize applications 

### ISSUE THREAD

**Issue by Benjamin-Philip** [hexpm/hexpm#1030]
Currently to create a key, one would send a POST request with username and passwd to `api/keys`.

However, to support Web Auth #1024, I plan to implement the [device flow](https://datatracker.ietf.org/doc/html/rfc8628)
which will require **different** endpoints (all of which are POST) for the following:

- Device authorization request and response
- User Interaction (user code entry)
- Device Access Token Request

If I were to have all these endpoints under `api/keys/*`, it would quickly become messy (in my opinion), and
would make it difficult (a better word would be inelegant) to add more OAuth grantypes (as there will be multiple
ways of generating a token)

That's why I propose to have a `api/oauth` scope (or anything of the same effect), and have the endpoints for
creating API keys (which is in essence oauth) under it. We could have an `api/oauth/passwd` for the password
grantype, `api/oauth/device/*`, for the above mentioned actions.

However, the `api/oauth` scope will **only** be used to **create** tokens; reading, updating, and deleting will remain
under the `api/keys` scope.

**Comment by Benjamin-Philip** [comment]
@wojtekmach , any thoughts ?

**Comment by ericmj** [comment]
My initial thought would be that oauth tokens and API keys should be separate entities, at least to the outside. So I agree that new endpoints are needed under `api/oauth` but I am not sure if reading, updating and deleting oauth tokens should be under `api/keys`. I think those actions should also be under `api/oauth`.

**Comment by Benjamin-Philip** [comment]
> I am not sure if reading, updating and deleting oauth tokens should be under api/keys. I think those actions should also be under api/oauth.

My general logic is that oauth does not specify the *management* (i.e updating and deleteing) of tokens (atleast to my knowledge), so we would manage
the tokens under `api/keys`. Also, I am refering to Hex api keys as a *token*, so if you think setting up JWT is worth the effort, I would be more
than happy to do so.

**Comment by ericmj** [comment]
I am not sure API keys and oauth tokens should be the same entities. In the future we may want to deprecate API keys in favor of oauth tokens or somehow roll them into oauth tokens, but right now that's not really clear so should they share the same API endpoints?

It's hard to comment on the location of endpoints before seeing the proposed API. When we know what the API will be or if you already have an idea and can outline it, it will be easier to discuss around that.

**Comment by Benjamin-Philip** [comment]
I'm of the opinion that we should switch to using a standardized token specification like JWT or SAML.

However, I haven't got a "complete" api in mind; I was thinking in terms of a device flow, and didn't think in terms of changing the token...

Do note, if any one else has got any ideas, please feel to suggest them here.

**Comment by ericmj** [comment]
> I'm of the opinion that we should switch to using a standardized token specification like JWT or SAML.

What would the benefits of this be, would it help us implementing web auth? I have nothing against this change but would like to understand the benefits and drawbacks to help us prioritize.

> assuming we setup OAuth for hexpm, would we allow third party clients to register ?

Yes, I would say so. If we already support oauth we can add support for third parties to use it but unless it's little extra work I would wait until we have implemented web auth.

Currently we are discussing three things: 1. Add web auth through oauth device authorization. 2. Changing the token specification 3. Add third party registration support.

I think all could be great improvements but makes the scope quite big so I think we should focus on web auth unless they are required to implement web auth or is more convenient to implement before. WDYT?

**Comment by Benjamin-Philip** [comment]
@ericmj , 

> What would the benefits of this be, would it help us implementing web auth? I have nothing against this change but would like to understand the benefits and drawbacks to help us prioritize.

*What* the token is doesn't make a difference to the web auth flow. I want to use a standardized token *because* it's standardized. With us having a standardized token, we don't need to worry about security caveats, it might make working the hexpm api easier, and most of all since it's standardized,  there are well defined protocols for creating tokens, managing headers etc when we use a token specification, which we don't need to design for the current API keys. Do note that [jwt](https://datatracker.ietf.org/doc/html/rfc7519) is considered a defacto standard for OAuth tokens.

> Yes, I would say so. If we already support oauth we can add support for third parties to use it but unless it's little extra work I would wait until we have implemented web auth.

I too think that we should support third party clients. I don't think it's going change the amount of work to do overall (though 
setting up a UI and all would prove tiresome)

> Currently we are discussing three things: 1. Add web auth through oauth device authorization. 2. Changing the token specification 3. Add third party registration support.

The web auth depends on some sort oauth device authorization, so I don't think there is going to be any change on that.
There are probably going to be more questions to come; 2 - 3 are just some regarding to how we would implement OAuth.
I take it that you are okay implementing OAuth to authorize clients ?

> I think all could be great improvements but makes the scope quite big so I think we should focus on web auth unless they are required to implement web auth or is more convenient to implement before. WDYT?

Having proper oauth support will certainly make things much easier for implementing web auth, but if we are to take this path,
it would involve a lot of yak shaving.

I think for now I will have the web auth endpoints at `api/keys/device/*` temporarily till we have oauth setup.  Any other suggestions ?

**Comment by wojtekmach** [comment]
I'm out this week so I won't be able to follow discussions but I'll be around after the weekend.

Off the top of head:

> `api/keys/device/*`

`/api` endpoints requires token authentication and `/dashboard` etc endpoints require cookie authentication, and for the web auth flow we need the latter. I am not sure we should start requiring cookie auth in an /api endpoint, even as temporary measure.

Our `mix hex.user auth --web` would open up a browser window but later the server would somehow communicate back the generated token to the terminal, were you able to find out how GitHub and similar solutions are doing that?

> Do note that jwt is considered a defacto standard for OAuth tokens.

JWT support most likely would require a dependency and we (effectively) can't have dependencies in hex_core as it is meant to be vendored.

> I too think that we should support third party clients. I don't think it's going change the amount of work to do overall (though
**setting up a UI and all would prove tiresome**)

> Having proper oauth support will certainly make things much easier for implementing web auth, but if we are to take this path, **it would involve a lot of yak shaving**.

(emphasis mine)

Yeah I'm really concerned about the added effort and complexity.

What's the simplest possible thing that would work and that we could (but not necessarily will anytime soon cause we may not have availability etc) iterate on?

**Comment by Benjamin-Philip** [comment]
> Our mix hex.user auth --web would open up a browser window but later the server would somehow communicate back the generated token to the terminal, were you able to find out how GitHub and similar solutions are doing that?

See https://docs.github.com/en/free-pro-team@latest/developers/apps/authorizing-oauth-apps#device-flow.

It's also clearly described in the RFC document above.

**Comment by Benjamin-Philip** [comment]
> JWT support most likely would require a dependency and we (effectively) can't have dependencies in hex_core as it is meant to be vendored.

I'm not sure about that. As far as I know, the client doesn't need to make changes to token, but just save it.

Would `login/device` do for now ?

**Comment by josevalim** [comment]
OAuth is meant to handle a much wider and more complex variety of problems than web auth. I think it is worth taking a step back and evaluate if there are alternatives or if there is a subset of problems that could be implemented without having to adhere to the whole OAuth specification.

**Comment by Benjamin-Philip** [comment]
I think that's what I'll end up doing for web auth.

Though I think generally it is worth adhering to OAuth. Maybe we should work on this after implementing web auth ?

**Comment by josevalim** [comment]
> Though I think generally it is worth adhering to OAuth.

Which purpose do you have in mind? I don't think anyone is planning to delegate authentication to Hex. And for the use case described here, we can do simpler than OAuth. :)

**Comment by Benjamin-Philip** [comment]
I don't have much on my mind. It just "felt" like the right thing to do.. Though I feel setting this up would eliminate any security concerns that may crop up.

I looked at pypi and rubygems and they don't  have anything of the lines too, so I think we should only work on this if we find a good use case some time in the future.

Also any comments on `login/device` ?
I plan to work on it this Friday, and it would not be nice if we find some better path after I finished working on this.
